Client/Frontend/Home/HistoryHighlights/HistoryHighlightsManager.swift

```diff
@@ -0,0 +1,125 @@
+                                         and tabs: [Tab],
+                                         shouldGroupHighlights: Bool = false,
+                                         completion: @escaping ([HighlightItem]?) -> Void) {
+        HistoryHighlightsManager.shouldGroupHighlights = shouldGroupHighlights
+        fetchHighlights(with: profile) { highlights in
+
+            guard let highlights = highlights, !highlights.isEmpty else {
+                completion(nil)
+                return
+            }
+
+            let filterHighlights = highlights.filter { highlights in
+                !tabs.contains { highlights.urlFromString == $0.url }
+            }
+
+            if shouldGroupHighlights {
+                buildSearchGroups(with: profile, and: filterHighlights) { groups, filterHighlights in
+
+                    let collatedHighlights = collateForRecentlySaved(from: groups, and: filterHighlights)
+                    completion(Array(collatedHighlights.prefix(9)))
+                }
+            } else {
+                completion(Array(filterHighlights.prefix(9)))
+            }
+        }
+    }
+
+    // MARK: - Data fetching functions
+
+    private static func fetchHighlights(with profile: Profile,
+                                        andLimit limit: Int32 = 1000,
+                                        completion: @escaping ([HistoryHighlight]?) -> Void) {
+
+        profile.places.getHighlights(weights: HistoryHighlightWeights(viewTime: self.defaultViewTimeWeight,
+                                                                      frequency: self.defaultFrequencyWeight),
+                                     limit: limit).uponQueue(.main) { result in
+
+            guard let ASHighlights = result.successValue, !ASHighlights.isEmpty else { return completion(nil) }
+
+            completion(ASHighlights)
+        }
+    }
+
+    // MARK: - Helper functions
+
+    private static func buildSearchGroups(with profile: Profile,
+                                          and highlights: [HistoryHighlight],
+                                          completion: @escaping ([ASGroup<HistoryHighlight>]?, [HistoryHighlight]) -> Void) {
+        SearchTermGroupsManager.getHighlightGroups(with: profile,
+                                                   from: highlights,
+                                                   using: .orderedAscending) { groups, filteredItems in
+            completion(groups, filteredItems)
+        }
+    }
+
+
+    /// Collate a `HistoryHighlight` group and individual `HistoryHighlight`, the result array alternates them starting with individual highlights.
+    ///  Because groups could be nil, the `HighlightItem` array get initialized with the `HistoryHighlight` array and if groups are not nil are inserted in the odd index of the array.
+    ///  In case the individual items are done, the rest of the group array gets appended to the result array
```

Comment:
```suggestion
    /// Collate `HistoryHighlight` groups and individual `HistoryHighlight` items, such that
    /// the resulting array alternates between them, starting with individual highlights.
    /// Because groups could be nil, the `HighlightItem` array gets initialized with the 
    /// `HistoryHighlight` array and, if not `nil`, groups are then inserted in the odd index
    /// of the array. In case the individual items are done, the rest of the group array gets 
    /// appended to the result array.
    ///
```